src/sema/mutability.rs

```diff
@@ -307,17 +358,27 @@ fn read_expression(expr: &Expression, state: &mut StateCheck) -> bool {
         | Expression::PostIncrement { expr, .. }
         | Expression::PostDecrement { expr, .. } => {
             expr.recurse(state, write_expression);
+            return false;
         }
         Expression::Assign { left, right, .. } => {
             right.recurse(state, read_expression);
             left.recurse(state, write_expression);
+            return false;
+        }
+        Expression::StorageVariable { loc, .. } => {
+            state.data_account |= DataAccountUsage::READ;
+            state.read(loc);
+            return false;
         }
-        Expression::StorageVariable { loc, .. } => state.read(loc),
         Expression::StorageArrayLength { loc, .. } | Expression::StorageLoad { loc, .. } => {
-            state.read(loc)
+            state.data_account |= DataAccountUsage::READ;
+            state.read(loc);
+            return false;
         }
         Expression::Subscript { loc, array_ty, .. } if array_ty.is_contract_storage() => {
-            state.read(loc)
+            state.data_account |= DataAccountUsage::READ;
+            state.read(loc);
+            return false;
```

Comment:
If we continue recursing, there would be two mutability errors. One for the subscript and other for the storage variable nested within the subscript expression. Perhaps the best solution here is just to track the variable access instead of the subscript.

Reply:
> Perhaps the best solution here is just to track the variable access instead of the subscript.

There is a test case that fails if I do that. I need to keep the expressions separated and not recurse on on them.

Reply:
It's fundamentally broken. 

```solidity
contract C {
	bool x;
	int[4] a;

	function test() public view returns (int) {
		return foo()[1];
	}

	function foo() internal returns (int[4] storage) {
		x = true;
		return a;
	}

}
```

Reply:
Please, do not say it is broken without testing. The IDL generated for this contract is correct:
```json
    {
      "name": "test",
      "accounts": [
        {
          "name": "dataAccount",
          "isMut": true,
          "isSigner": false,
          "isOptional": false
        }
      ],
      "args": [],
      "returns": "i256"
    }
```

The collection of accounts between function calls happens in codegen.

Reply:
`account_collection.rs`

Reply:
> > Please, do not say it is broken without testing. The IDL generated for this contract is correct:
> 
> It is broken and this example shows exactly why. This example should not compile: a view function calls a default non-view function (which mutates the state). The same solidity does not compile with solc.

I got what you mean. Keep in mind that this is also broken in the main branch. I did not introduce this problem.

Reply:
Of course, we need to fix it. When reviewing, please be clear whether the PR introduced the regression or it is something broken in the main that you've only figured out now.

Reply:
Ok, maybe this particular case was also broken in the main branch. 

The point is this: you can't just not traverse parts of the ast when looking for mutability.

Reply:
> The point is this: you can't just not traverse parts of the ast when looking for mutability.

Right now, the main branch stops traversing when we match one of the cases in `fn read_expression`.
https://github.com/hyperledger/solang/blob/90c687ecb52dd4c75adda3c52443c0b5725ab16c/src/sema/mutability.rs#L403

Recursing down everything risks adding duplicate error messages about mutability, as in this example:
```solidity
        function x(int[] storage x) pure returns (int) { return x[1]; }
```

The deeper you go, the less context you have. When we are simply forwarding storage pointers, we cannot raise an error when we reach `Expression::Variable`: https://github.com/hyperledger/solang/blob/main/tests/contract_testcases/solana/mutability.sol

Mutability checks would be particularly easier if done in codegen. We already have `Instr::StorageLoad` and `Instr::StorageStore` that would indicate whether we read or write the state variables.

If the mutability check needs a major overhaul, I prefer removing the data account detection from there and doing it in `account_collection.rs` so we can fix all the problems later.

Reply:
My goal here was not to make mutability check perfect, but make it work for the tests we have while collecting the data account. Not continuing the traversal was an existing problem and its solution is not simple to include in this PR. I am happy to work on this separately.

Reply:
> > The point is this: you can't just not traverse parts of the ast when looking for mutability.
> 
> Right now, the main branch stops traversing when we match one of the cases in `fn read_expression`.

I just replicated the main branch's behavior by stopping.

Reply:
If you want to work on this on separate PR, sure.

Reply:
> If you want to work on this on separate PR, sure.

Thanks! I should make a start on this today or tomorrow. As the problem also exists in the main branch, I believe it deserves a separate piece of work.

Reply:
I'll track this issue at #1493.